### ISSUE TITLE
Loading patches from large-scale point clouds with DataSet and DataLoader

### ISSUE THREAD

**Issue by MiZhenxing** [pyg-team/pytorch_geometric#841]
## ❓ Questions & Help

Hello, thanks for your amazing job.

I am trying to reimplement the tangent convolution [https://github.com/tatarchm/tangent_conv](url) with pytorch_geometric. The tangent convolution allows subdividing the point clouds using overlapping bounding boxes and processing them individually. I have questions on how to load data efficiently.

To make questions clearer, I give an example below.

I have a dataset of 50 point clouds. Each of them has millions of points. 
I want to load one point cloud into the memory and subdividing them with bounding boxes into patches. Then I feed each patch into the network for  training. After all patches are processed, the next point cloud will be loaded.

```
// the pseudocode, pc means point cloud
for pc_file_name in pc_file_names:
   pc_data = load_pc(pc_file_name ) // load a point cloud
   patch_list = subdivide(pc_data) // subdivide a point cloud into overlapping patches
   for patch_data in patch_list:
       pred = tangent_conv(patch_data) // process one patch
```
My question is whether I can implement a unified DataSet and DataLoader interface that load one pathch a time like code below and in the mean time surport data prefetching and multi-gpu processing?

```
// the pseudocode
def __getitem__(self, idx): // idx is a global idx for one patch
    pc_idx, local_patch_idx = transform_idx(idx) // transform the global idx into pc idx and local patch idx
    if not in_memory(pc_idx):
        pc_data = load_pc(pc_idx) // if pc of pc_idx is not in memory, load it
        patch_list  = subdivide(pc_data) // subdivide the newly loaded pc_data
    return patch_list[patch_idx] // get a patch of data
```
Sorry for my unclear English. When processing large-scale point clouds, subdividing is an important solution and I think it is worth of more investigation.  Thank you very much!

**Comment by rusty1s** [comment]
Interesting question! What prevents you from implementing your pseudo-code? This should work like a charm provided that you can hold a single point cloud in memory as long as possible.

**Comment by MiZhenxing** [comment]
Hi, I am trying working on it. I think I may face synchronization problems when loading and subdividing multiple point clouds with one DataSet object. But it may also not appear. I will give more feedbacks during my implementation. Thank you.

**Comment by rusty1s** [comment]
Ah, I see. You may need to add a lock to load and subdivide your data.

**Comment by MiZhenxing** [comment]
Yes, I want to first load a point cloud. Then when I need a patch, The DataSet object extracts it from the point cloud. When all  patches are processed, the DataSet object will remove the point cloud from memory and load another point cloud. So if the DataSet prefetches a patch, the corresponding point cloud should be guaranteed in the memory, may be with a lock. This may cause problems when I want to asynchronously load data in order to speed up the training.
